{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### _Anton Lavrinienko_\n",
    "\n",
    "# Validation experiment data analysis\n",
    "\n",
    "### Analysis Notebook\n",
    "\n",
    "This notebook contains code that was used to process the raw read data from the validation experiment.\n",
    "\n",
    "### Requirements\n",
    "\n",
    "qiime2 is required to run most of the commands described below (i.e., qiime2-amplicon-2024.2 or qiime2-amplicon-2024.10 release). To install qiime2 follow these instructions: https://docs.qiime2.org/2024.2/install/. To activate the qiime2 conda environment run: conda activate qiime2-amplicon-2024.2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check the path and content\n",
    "!pwd\n",
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the validation experiment data (n=276 samples: 15 subjects x 8+8 treatment/storage + 1 fresh + 1 directly-plated sample, and 6 negative controls) according to sequencing runs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir demux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported manifest-val-exp-batch1-202412.txt as PairedEndFastqManifestPhred33V2 to demux/val-exp-batch1-demux.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: demux/val-exp-batch1-demux.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#import BATCH1 data into qiime2\n",
    "!qiime tools import \\\n",
    "--type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "--input-path manifest-val-exp-batch1-202412.txt \\\n",
    "--output-path demux/val-exp-batch1-demux.qza \\\n",
    "--input-format PairedEndFastqManifestPhred33V2\n",
    "\n",
    "#run qiime2 summary for BATCH1 data\n",
    "!qiime demux summarize \\\n",
    "--i-data demux/val-exp-batch1-demux.qza \\\n",
    "--o-visualization demux/val-exp-batch1-demux.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported manifest-val-exp-batch2-202412.txt as PairedEndFastqManifestPhred33V2 to demux/val-exp-batch2-demux.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: demux/val-exp-batch2-demux.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mImported manifest-val-exp-batch3-202412.txt as PairedEndFastqManifestPhred33V2 to demux/val-exp-batch3-demux.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: demux/val-exp-batch3-demux.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mImported manifest-val-exp-batch4-202412.txt as PairedEndFastqManifestPhred33V2 to demux/val-exp-batch4-demux.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: demux/val-exp-batch4-demux.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#import BATCH2 data into qiime2\n",
    "!qiime tools import \\\n",
    "--type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "--input-path manifest-val-exp-batch2-202412.txt \\\n",
    "--output-path demux/val-exp-batch2-demux.qza \\\n",
    "--input-format PairedEndFastqManifestPhred33V2\n",
    "\n",
    "#run qiime2 summary for BATCH2 data\n",
    "!qiime demux summarize \\\n",
    "--i-data demux/val-exp-batch2-demux.qza \\\n",
    "--o-visualization demux/val-exp-batch2-demux.qzv\n",
    "\n",
    "\n",
    "#import BATCH3 data into qiime2\n",
    "!qiime tools import \\\n",
    "--type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "--input-path manifest-val-exp-batch3-202412.txt \\\n",
    "--output-path demux/val-exp-batch3-demux.qza \\\n",
    "--input-format PairedEndFastqManifestPhred33V2\n",
    "\n",
    "#run qiime2 summary for BATCH3 data\n",
    "!qiime demux summarize \\\n",
    "--i-data demux/val-exp-batch3-demux.qza \\\n",
    "--o-visualization demux/val-exp-batch3-demux.qzv\n",
    "\n",
    "\n",
    "#import BATCH4 data into qiime2\n",
    "!qiime tools import \\\n",
    "--type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "--input-path manifest-val-exp-batch4-202412.txt \\\n",
    "--output-path demux/val-exp-batch4-demux.qza \\\n",
    "--input-format PairedEndFastqManifestPhred33V2\n",
    "\n",
    "#run qiime2 summary for BATCH4 data\n",
    "!qiime demux summarize \\\n",
    "--i-data demux/val-exp-batch4-demux.qza \\\n",
    "--o-visualization demux/val-exp-batch4-demux.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data denoising"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Denoise the read data to identify and catalogue amplicon sequence variants (ASVs) using dada2 separately for each sequencing run. Then, use the output data to assess the denoising performance by summarizing dada2 stats and the feature data. At this step, we truncate forward reads at 200bp and reverse reads at 160bp (with these settings, the median quality scores for the remaining reads will hover above >Q35)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir dada2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Note that this code chunk was launched using sbatch job on Euler (HPC) as it would struggle to run interactively via jupyter notebook.\n",
    "\n",
    "#BATCH1 data denoising\n",
    "qiime dada2 denoise-paired \\\n",
    "    --i-demultiplexed-seqs demux/val-exp-batch1-demux.qza \\\n",
    "    --p-trunc-len-f 200 \\\n",
    "    --p-trunc-len-r 160 \\\n",
    "    --p-n-threads 0 \\\n",
    "    --o-table dada2/val-exp-batch1-feature-table.qza \\\n",
    "    --o-representative-sequences dada2/val-exp-batch1-rep-seqs.qza \\\n",
    "    --o-denoising-stats dada2/val-exp-batch1-denoising-stats.qza \\\n",
    "    --verbose\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file dada2/val-exp-batch1-denoising-stats.qza \\\n",
    "    --o-visualization dada2/val-exp-batch1-denoising-stats.qzv\n",
    "\n",
    "qiime feature-table summarize \\\n",
    "    --i-table dada2/val-exp-batch1-feature-table.qza \\\n",
    "    --o-visualization dada2/val-exp-batch1-feature-table.qzv\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "    --i-data dada2/val-exp-batch1-rep-seqs.qza \\\n",
    "    --o-visualization dada2/val-exp-batch1-rep-seqs.qzv\n",
    "\n",
    "#BATCH2 data denoising\n",
    "qiime dada2 denoise-paired \\\n",
    "    --i-demultiplexed-seqs demux/val-exp-batch2-demux.qza \\\n",
    "    --p-trunc-len-f 200 \\\n",
    "    --p-trunc-len-r 160 \\\n",
    "    --p-n-threads 0 \\\n",
    "    --o-table dada2/val-exp-batch2-feature-table.qza \\\n",
    "    --o-representative-sequences dada2/val-exp-batch2-rep-seqs.qza \\\n",
    "    --o-denoising-stats dada2/val-exp-batch2-denoising-stats.qza \\\n",
    "    --verbose\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file dada2/val-exp-batch2-denoising-stats.qza \\\n",
    "    --o-visualization dada2/val-exp-batch2-denoising-stats.qzv\n",
    "\n",
    "qiime feature-table summarize \\\n",
    "    --i-table dada2/val-exp-batch2-feature-table.qza \\\n",
    "    --o-visualization dada2/val-exp-batch2-feature-table.qzv\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "    --i-data dada2/val-exp-batch2-rep-seqs.qza \\\n",
    "    --o-visualization dada2/val-exp-batch2-rep-seqs.qzv\n",
    "\n",
    "#BATCH3 data denoising\n",
    "qiime dada2 denoise-paired \\\n",
    "    --i-demultiplexed-seqs demux/val-exp-batch3-demux.qza \\\n",
    "    --p-trunc-len-f 200 \\\n",
    "    --p-trunc-len-r 160 \\\n",
    "    --p-n-threads 0 \\\n",
    "    --o-table dada2/val-exp-batch3-feature-table.qza \\\n",
    "    --o-representative-sequences dada2/val-exp-batch3-rep-seqs.qza \\\n",
    "    --o-denoising-stats dada2/val-exp-batch3-denoising-stats.qza \\\n",
    "    --verbose\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file dada2/val-exp-batch3-denoising-stats.qza \\\n",
    "    --o-visualization dada2/val-exp-batch3-denoising-stats.qzv\n",
    "\n",
    "qiime feature-table summarize \\\n",
    "    --i-table dada2/val-exp-batch3-feature-table.qza \\\n",
    "    --o-visualization dada2/val-exp-batch3-feature-table.qzv\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "    --i-data dada2/val-exp-batch3-rep-seqs.qza \\\n",
    "    --o-visualization dada2/val-exp-batch3-rep-seqs.qzv\n",
    "\n",
    "#BATCH4 data denoising\n",
    "qiime dada2 denoise-paired \\\n",
    "    --i-demultiplexed-seqs demux/val-exp-batch4-demux.qza \\\n",
    "    --p-trunc-len-f 200 \\\n",
    "    --p-trunc-len-r 160 \\\n",
    "    --p-n-threads 0 \\\n",
    "    --o-table dada2/val-exp-batch4-feature-table.qza \\\n",
    "    --o-representative-sequences dada2/val-exp-batch4-rep-seqs.qza \\\n",
    "    --o-denoising-stats dada2/val-exp-batch4-denoising-stats.qza \\\n",
    "    --verbose\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file dada2/val-exp-batch4-denoising-stats.qza \\\n",
    "    --o-visualization dada2/val-exp-batch4-denoising-stats.qzv\n",
    "\n",
    "qiime feature-table summarize \\\n",
    "    --i-table dada2/val-exp-batch4-feature-table.qza \\\n",
    "    --o-visualization dada2/val-exp-batch4-feature-table.qzv\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "    --i-data dada2/val-exp-batch4-rep-seqs.qza \\\n",
    "    --o-visualization dada2/val-exp-batch4-rep-seqs.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data cleanup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the qiime2 decontam plugin in qiime2-amplicon-2024.10 per sequencing batch to identify and remove potential contaminants and spurious ASVs. Use the prevalence method in decontam to cleanup feature data using the distribution of ASVs from the technical control samples (per sequencing batch; see more information here: https://jordenrabasco.github.io/Q2_Decontam_Tutorial.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[DecontamScore] to: dada2/decontam-prev-scores-val-exp-batch1.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/decontam-prev-scores-viz-val-exp-batch1.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#Note that qiime2-amplicon-2024.10 was used to run this and also other decontam-related code chunks below.\n",
    "\n",
    "#BATCH1 data\n",
    "!qiime quality-control decontam-identify \\\n",
    "--i-table dada2/val-exp-batch1-feature-table.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-method prevalence \\\n",
    "--p-prev-control-column SampleType \\\n",
    "--p-prev-control-indicator TechControl \\\n",
    "--o-decontam-scores dada2/decontam-prev-scores-val-exp-batch1.qza\n",
    "\n",
    "!qiime quality-control decontam-score-viz \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch1.qza \\\n",
    "--i-table dada2/val-exp-batch1-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch1-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-visualization dada2/decontam-prev-scores-viz-val-exp-batch1.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine the decontam output generated above and remove features identified as potential contaminants from the feature table and the representative sequences per batch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch1-feature-table-decont.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch1-rep-seqs-decont.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch1-feature-table-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch1-rep-seqs-decont.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#remove contaminants and spurious ASVs identified by decontam\n",
    "#BATCH1 data\n",
    "!qiime quality-control decontam-remove \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch1.qza \\\n",
    "--i-table dada2/val-exp-batch1-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch1-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-filtered-table dada2/val-exp-batch1-feature-table-decont.qza \\\n",
    "--o-filtered-rep-seqs dada2/val-exp-batch1-rep-seqs-decont.qza\n",
    "\n",
    "#visualize decontaminated feature data\n",
    "#BATCH1 data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch1-feature-table-decont.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch1-rep-seqs-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch1-rep-seqs-decont.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run decontam on the rest of the validation experiment data per sequencing batch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[DecontamScore] to: dada2/decontam-prev-scores-val-exp-batch2.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/decontam-prev-scores-viz-val-exp-batch2.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[DecontamScore] to: dada2/decontam-prev-scores-val-exp-batch3.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/decontam-prev-scores-viz-val-exp-batch3.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[DecontamScore] to: dada2/decontam-prev-scores-val-exp-batch4.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/decontam-prev-scores-viz-val-exp-batch4.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#BATCH2 data\n",
    "!qiime quality-control decontam-identify \\\n",
    "--i-table dada2/val-exp-batch2-feature-table.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-method prevalence \\\n",
    "--p-prev-control-column SampleType \\\n",
    "--p-prev-control-indicator TechControl \\\n",
    "--o-decontam-scores dada2/decontam-prev-scores-val-exp-batch2.qza\n",
    "\n",
    "!qiime quality-control decontam-score-viz \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch2.qza \\\n",
    "--i-table dada2/val-exp-batch2-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch2-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-visualization dada2/decontam-prev-scores-viz-val-exp-batch2.qzv\n",
    "\n",
    "#BATCH3 data\n",
    "!qiime quality-control decontam-identify \\\n",
    "--i-table dada2/val-exp-batch3-feature-table.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-method prevalence \\\n",
    "--p-prev-control-column SampleType \\\n",
    "--p-prev-control-indicator TechControl \\\n",
    "--o-decontam-scores dada2/decontam-prev-scores-val-exp-batch3.qza\n",
    "\n",
    "!qiime quality-control decontam-score-viz \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch3.qza \\\n",
    "--i-table dada2/val-exp-batch3-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch3-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-visualization dada2/decontam-prev-scores-viz-val-exp-batch3.qzv\n",
    "\n",
    "#BATCH4 data\n",
    "!qiime quality-control decontam-identify \\\n",
    "--i-table dada2/val-exp-batch4-feature-table.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-method prevalence \\\n",
    "--p-prev-control-column SampleType \\\n",
    "--p-prev-control-indicator TechControl \\\n",
    "--o-decontam-scores dada2/decontam-prev-scores-val-exp-batch4.qza\n",
    "\n",
    "!qiime quality-control decontam-score-viz \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch4.qza \\\n",
    "--i-table dada2/val-exp-batch4-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch4-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-visualization dada2/decontam-prev-scores-viz-val-exp-batch4.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch2-feature-table-decont.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch2-rep-seqs-decont.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch2-feature-table-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch2-rep-seqs-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch3-feature-table-decont.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch3-rep-seqs-decont.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch3-feature-table-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch3-rep-seqs-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch4-feature-table-decont.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch4-rep-seqs-decont.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch4-feature-table-decont.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch4-rep-seqs-decont.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#remove contaminants and spurious ASVs identified by decontam\n",
    "#BATCH2 data\n",
    "!qiime quality-control decontam-remove \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch2.qza \\\n",
    "--i-table dada2/val-exp-batch2-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch2-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-filtered-table dada2/val-exp-batch2-feature-table-decont.qza \\\n",
    "--o-filtered-rep-seqs dada2/val-exp-batch2-rep-seqs-decont.qza\n",
    "\n",
    "#visualize decontaminated feature data\n",
    "#BATCH2 data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch2-feature-table-decont.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch2-rep-seqs-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch2-rep-seqs-decont.qzv\n",
    "\n",
    "\n",
    "\n",
    "#remove contaminants and spurious ASVs identified by decontam\n",
    "#BATCH3 data\n",
    "!qiime quality-control decontam-remove \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch3.qza \\\n",
    "--i-table dada2/val-exp-batch3-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch3-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-filtered-table dada2/val-exp-batch3-feature-table-decont.qza \\\n",
    "--o-filtered-rep-seqs dada2/val-exp-batch3-rep-seqs-decont.qza\n",
    "\n",
    "#visualize decontaminated feature data\n",
    "#BATCH3 data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch3-feature-table-decont.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch3-rep-seqs-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch3-rep-seqs-decont.qzv\n",
    "\n",
    "\n",
    "\n",
    "#remove contaminants and spurious ASVs identified by decontam\n",
    "#BATCH4 data\n",
    "!qiime quality-control decontam-remove \\\n",
    "--i-decontam-scores dada2/decontam-prev-scores-val-exp-batch4.qza \\\n",
    "--i-table dada2/val-exp-batch4-feature-table.qza \\\n",
    "--i-rep-seqs dada2/val-exp-batch4-rep-seqs.qza \\\n",
    "--p-threshold 0.1 \\\n",
    "--o-filtered-table dada2/val-exp-batch4-feature-table-decont.qza \\\n",
    "--o-filtered-rep-seqs dada2/val-exp-batch4-rep-seqs-decont.qza\n",
    "\n",
    "#visualize decontaminated feature data\n",
    "#BATCH4 data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch4-feature-table-decont.qzv\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch4-rep-seqs-decont.qza \\\n",
    "--o-visualization dada2/val-exp-batch4-rep-seqs-decont.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter technical controls from the feature tables for each sequencing batch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch1-feature-table-decont-negfilt.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch1-feature-table-decont-negfilt.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#BATCH1\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-where \"[SampleType]='TechControl'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/val-exp-batch1-feature-table-decont-negfilt.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont-negfilt.qza \\\n",
    "--o-visualization dada2/val-exp-batch1-feature-table-decont-negfilt.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch2-feature-table-decont-negfilt.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch2-feature-table-decont-negfilt.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch3-feature-table-decont-negfilt.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch3-feature-table-decont-negfilt.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch4-feature-table-decont-negfilt.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch4-feature-table-decont-negfilt.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#BATCH2\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-where \"[SampleType]='TechControl'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/val-exp-batch2-feature-table-decont-negfilt.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont-negfilt.qza \\\n",
    "--o-visualization dada2/val-exp-batch2-feature-table-decont-negfilt.qzv\n",
    "\n",
    "\n",
    "#BATCH3\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-where \"[SampleType]='TechControl'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/val-exp-batch3-feature-table-decont-negfilt.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont-negfilt.qza \\\n",
    "--o-visualization dada2/val-exp-batch3-feature-table-decont-negfilt.qzv\n",
    "\n",
    "\n",
    "#BATCH4\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--p-where \"[SampleType]='TechControl'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/val-exp-batch4-feature-table-decont-negfilt.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont-negfilt.qza \\\n",
    "--o-visualization dada2/val-exp-batch4-feature-table-decont-negfilt.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the feature data based on frequency. At this step, we remove ASVs represented with less than 10 reads across all samples in each sequencing batch separately. This is because such rare ASVs are more likely to originate from sequencing errors, and they are less useful for comparisons also because in most cases they are found in a single sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#BATCH1\n",
    "#filter low-frequency and low-prevalence features from the feature table\n",
    "!qiime feature-table filter-features \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont-negfilt.qza \\\n",
    "--p-min-frequency 10 \\\n",
    "--o-filtered-table dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qzv\n",
    "\n",
    "#filter features from representative sequences to match observations to the final filtered feature table above\n",
    "!qiime feature-table filter-seqs \\\n",
    "--i-data dada2/val-exp-batch1-rep-seqs-decont.qza \\\n",
    "--i-table dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-filtered-data dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#BATCH2\n",
    "#filter low-frequency and low-prevalence features from the feature table\n",
    "!qiime feature-table filter-features \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont-negfilt.qza \\\n",
    "--p-min-frequency 10 \\\n",
    "--o-filtered-table dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qzv\n",
    "\n",
    "#filter features from representative sequences to match observations to the final filtered feature table above\n",
    "!qiime feature-table filter-seqs \\\n",
    "--i-data dada2/val-exp-batch2-rep-seqs-decont.qza \\\n",
    "--i-table dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-filtered-data dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qzv\n",
    "\n",
    "\n",
    "#BATCH3\n",
    "#filter low-frequency and low-prevalence features from the feature table\n",
    "!qiime feature-table filter-features \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont-negfilt.qza \\\n",
    "--p-min-frequency 10 \\\n",
    "--o-filtered-table dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qzv\n",
    "\n",
    "#filter features from representative sequences to match observations to the final filtered feature table above\n",
    "!qiime feature-table filter-seqs \\\n",
    "--i-data dada2/val-exp-batch3-rep-seqs-decont.qza \\\n",
    "--i-table dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-filtered-data dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qzv\n",
    "\n",
    "\n",
    "#BATCH4\n",
    "#filter low-frequency and low-prevalence features from the feature table\n",
    "!qiime feature-table filter-features \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont-negfilt.qza \\\n",
    "--p-min-frequency 10 \\\n",
    "--o-filtered-table dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qzv\n",
    "\n",
    "#filter features from representative sequences to match observations to the final filtered feature table above\n",
    "!qiime feature-table filter-seqs \\\n",
    "--i-data dada2/val-exp-batch4-rep-seqs-decont.qza \\\n",
    "--i-table dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qza \\\n",
    "--o-filtered-data dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--o-visualization dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the feature data are properly cleaned-up, we can merge feature data from each sequencing batch into a single feature table and representative sequences file - sanity-checked AL - OK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/merged-clean-feature-table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/merged-clean-feature-table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/merged-clean-rep-seqs.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/merged-clean-rep-seqs.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#merge all feature tables\n",
    "!qiime feature-table merge \\\n",
    "--i-tables dada2/val-exp-batch1-feature-table-decont-negfilt-filt10.qza \\\n",
    "--i-tables dada2/val-exp-batch2-feature-table-decont-negfilt-filt10.qza \\\n",
    "--i-tables dada2/val-exp-batch3-feature-table-decont-negfilt-filt10.qza \\\n",
    "--i-tables dada2/val-exp-batch4-feature-table-decont-negfilt-filt10.qza \\\n",
    "--verbose \\\n",
    "--o-merged-table dada2/merged-clean-feature-table.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/merged-clean-feature-table.qza \\\n",
    "--m-sample-metadata-file metadata-202412.txt \\\n",
    "--o-visualization dada2/merged-clean-feature-table.qzv\n",
    "\n",
    "#merge all representative sequences\n",
    "!qiime feature-table merge-seqs \\\n",
    "--i-data dada2/val-exp-batch1-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--i-data dada2/val-exp-batch2-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--i-data dada2/val-exp-batch3-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--i-data dada2/val-exp-batch4-rep-seqs-decont-negfilt-filt10.qza \\\n",
    "--verbose \\\n",
    "--o-merged-data dada2/merged-clean-rep-seqs.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/merged-clean-rep-seqs.qza \\\n",
    "--o-visualization dada2/merged-clean-rep-seqs.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign taxonomy to the feature data, and then filter the data based on taxonomic annotations (e.g., non-target ASVs such as chloroplast, mitochondria).\n",
    "\n",
    "Download the pre-trained Naive Bayes taxonomy classifier trained on the SILVA reference data from the qiime2 docs: https://docs.qiime2.org/2024.2/data-resources/ (in this analysis step we use qiime2-amplicon-2024.2), and check that the MD5 is matching the original file (i.e., Silva 138 99% OTUs from 515F/806R region of sequences (MD5: e05afad0fe87542704be96ff483824d4) sanity-checked AL - OK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "e05afad0fe87542704be96ff483824d4  /cluster/work/bokulich/alavrinienko/MV-project/SILVA-ref-data/silva-138-99-515-806-nb-classifier.qza\r\n"
     ]
    }
   ],
   "source": [
    "#check MD5\n",
    "!md5sum /cluster/work/bokulich/alavrinienko/MV-project/SILVA-ref-data/silva-138-99-515-806-nb-classifier.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Note that the code chunk below was launched as a sbatch job on Euler (HPC) as it would struggle to run interactively via jupyter notebook.\n",
    "\n",
    "#assign taxonomy\n",
    "qiime feature-classifier classify-sklearn \\\n",
    "    --i-classifier /cluster/work/bokulich/alavrinienko/MV-project/SILVA-ref-data/silva-138-99-515-806-nb-classifier.qza \\\n",
    "    --i-reads dada2/merged-clean-rep-seqs.qza \\\n",
    "    --o-classification dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "    --verbose\n",
    "\n",
    "#examine assigned taxonomy\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "    --o-visualization dada2/val-exp-merged-clean-silva-taxonomy.qzv\n",
    "\n",
    "#make a quick visual summary of the taxonomy composition across samples to identify potential contaminants and off-target reads\n",
    "qiime taxa barplot \\\n",
    "    --i-table dada2/merged-clean-feature-table.qza \\\n",
    "    --i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "    --m-metadata-file metadata-202412.txt \\\n",
    "    --o-visualization dada2/dirty-val-exp-merged-clean-silva-taxa-bar-plots.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter non-target ASVs, such as those assigned to chloroplast and mitochondria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/merged-clean-taxfilt-feature-table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/merged-clean-taxfilt-feature-table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureData[Sequence] to: dada2/merged-clean-taxfilt-rep-seqs.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/merged-clean-taxfilt-rep-seqs.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#filter ASVs assigned to mitochondria and chloroplast\n",
    "!qiime taxa filter-table \\\n",
    "--i-table dada2/merged-clean-feature-table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--p-exclude mitochondria,chloroplast \\\n",
    "--o-filtered-table dada2/merged-clean-taxfilt-feature-table.qza\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--m-sample-metadata-file metadata-202412.txt \\\n",
    "--o-visualization dada2/merged-clean-taxfilt-feature-table.qzv\n",
    "\n",
    "#filter features from representative sequences based on the filtered feature table above\n",
    "!qiime feature-table filter-seqs \\\n",
    "--i-data dada2/merged-clean-rep-seqs.qza \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--o-filtered-data dada2/merged-clean-taxfilt-rep-seqs.qza\n",
    "\n",
    "!qiime feature-table tabulate-seqs \\\n",
    "--i-data dada2/merged-clean-taxfilt-rep-seqs.qza \\\n",
    "--o-visualization dada2/merged-clean-taxfilt-rep-seqs.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validate taxonomy filtering and examine the microbiota composition across samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/clean-val-exp-merged-clean-taxfilt-silva-taxa-bar-plots.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#examine the taxonomy composition across samples using the merged, clean, and taxfilt filtered feature data\n",
    "!qiime taxa barplot \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--o-visualization dada2/clean-val-exp-merged-clean-taxfilt-silva-taxa-bar-plots.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylogeny and diversity analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construct a phylogenetic tree using FastTree, and use the longest branch to root the tree (‘midrooting’). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada2/aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada2/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: dada2/unrooted-tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: dada2/val-exp-rooted-tree.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "--i-sequences dada2/merged-clean-taxfilt-rep-seqs.qza \\\n",
    "--o-alignment dada2/aligned-rep-seqs.qza \\\n",
    "--o-masked-alignment dada2/masked-aligned-rep-seqs.qza \\\n",
    "--o-tree dada2/unrooted-tree.qza \\\n",
    "--o-rooted-tree dada2/val-exp-rooted-tree.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform alpha rarefaction analysis to examine the community coverage with the current sequencing effort. The median number of reads per sample in the final filtered feature data is 80177. This value will be used to generate alpha rarefaction curves to explore alpha diversity as a function of sequencing depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/val-exp-alpha-rarefaction.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-rarefaction \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--i-phylogeny dada2/val-exp-rooted-tree.qza \\\n",
    "--p-max-depth 80177 \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--o-visualization dada2/val-exp-alpha-rarefaction.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate various alpha- and beta-diversity metrics using qiime diversity plugin. At this step, we rarefy the data such that each sample has an equal sampling depth (number of reads) of 11987 reads. This is because the alpha rarefaction curves mostly level off at ~12K reads per sample, and the 11987 reads is the min number of reads per sample observed in the dataset (sample id P018-GPa-80C)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/core-metrics-results-rar11987/rarefied_table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: dada2/core-metrics-results-rar11987/faith_pd_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: dada2/core-metrics-results-rar11987/observed_features_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: dada2/core-metrics-results-rar11987/shannon_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: dada2/core-metrics-results-rar11987/evenness_vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: dada2/core-metrics-results-rar11987/unweighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: dada2/core-metrics-results-rar11987/weighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: dada2/core-metrics-results-rar11987/jaccard_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: dada2/core-metrics-results-rar11987/bray_curtis_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: dada2/core-metrics-results-rar11987/unweighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: dada2/core-metrics-results-rar11987/weighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: dada2/core-metrics-results-rar11987/jaccard_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: dada2/core-metrics-results-rar11987/bray_curtis_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/unweighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/weighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/jaccard_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/bray_curtis_emperor.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity core-metrics-phylogenetic \\\n",
    "--i-phylogeny dada2/val-exp-rooted-tree.qza \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--p-sampling-depth 11987 \\\n",
    "--m-metadata-file metadata-202412.txt \\\n",
    "--output-dir dada2/core-metrics-results-rar11987"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/rarefied-silva-taxa-bar-plots.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#examine the taxonomy composition across samples using rarefied feature data\n",
    "!qiime taxa barplot \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/rarefied-silva-taxa-bar-plots.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alpha diversity stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine differences in various alpha diversity metrics across samples based on the variables captured in the sample matadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/observed-features-group-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/faith-pd-group-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/shannon-group-significance.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#community richness\n",
    "!qiime diversity alpha-group-significance \\\n",
    "--i-alpha-diversity dada2/core-metrics-results-rar11987/observed_features_vector.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/observed-features-group-significance.qzv\n",
    "\n",
    "#faith's phylogenetic diversity\n",
    "!qiime diversity alpha-group-significance \\\n",
    "--i-alpha-diversity dada2/core-metrics-results-rar11987/faith_pd_vector.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/faith-pd-group-significance.qzv\n",
    "\n",
    "#shannon index\n",
    "!qiime diversity alpha-group-significance \\\n",
    "--i-alpha-diversity dada2/core-metrics-results-rar11987/shannon_vector.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/shannon-group-significance.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Beta diversity stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine differences in various beta diversity metrics across samples based on the Storage and Preservation metadata groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/bray_curtis-Storage-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/bray_curtis-Preservation-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/jaccard-Storage-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/jaccard-Preservation-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/unweighted-unifrac-Storage-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/unweighted-unifrac-Preservation-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/weighted-unifrac-Storage-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/weighted-unifrac-Preservation-significance.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#bray curtis\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/bray_curtis_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Storage \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/bray_curtis-Storage-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/bray_curtis_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Preservation \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/bray_curtis-Preservation-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "\n",
    "#jaccard\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/jaccard_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Storage \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/jaccard-Storage-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/jaccard_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Preservation \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/jaccard-Preservation-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "\n",
    "#unweighted unifrac\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/unweighted_unifrac_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Storage \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/unweighted-unifrac-Storage-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/unweighted_unifrac_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Preservation \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/unweighted-unifrac-Preservation-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "\n",
    "#weighted unifrac\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/weighted_unifrac_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Storage \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/weighted-unifrac-Storage-significance.qzv \\\n",
    "--p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "--i-distance-matrix dada2/core-metrics-results-rar11987/weighted_unifrac_distance_matrix.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Preservation \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/weighted-unifrac-Preservation-significance.qzv \\\n",
    "--p-pairwise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Machine learning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Classify samples and predict sample metadata values based on their microbial composition using machine learning algorithms with nested cross-validation (more details here: https://docs.qiime2.org/2024.10/tutorials/sample-classifier/). In this analysis, we use the rarefied feature data without samples that were sequenced fresh (i.e., only data from plated samples will be used for model trainign and testing)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/rarefied_table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#summarize rarefied data from the diversity core-metrics-phylogenetic pipeline\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_table.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/rarefied_table.qzv\n",
    "\n",
    "#filter samples that were sequenced fresh - retain only plated samples\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-where \"[State]='Fresh'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza\n",
    "\n",
    "#summarize and validate filtered rarefied data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Predict samples across different Preservation classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cluster/project/bokulich/alavrinienko/_conda_envs/qiime2-amplicon-2024.10/lib/python3.10/site-packages/q2_sample_classifier/utilities.py:715: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  imp = imp.groupby(imp.columns, axis=1).mean()\n",
      "Estimator Accuracy: 0.12549019607843137 ± 0.04223658672262357\n",
      "\u001b[32mSaved SampleData[ClassifierPredictions] to: dada2/core-metrics-results-rar11987/Preservation-predictions-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Importance] to: dada2/core-metrics-results-rar11987/Preservation-feature-importance-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[Probabilities] to: dada2/core-metrics-results-rar11987/Preservation-probabilities-RFncv.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#use filtered rarefied data for sample classifier\n",
    "#add --verbose flag to get mean accuracy ± SD for estimating variance in accuracy\n",
    "!qiime sample-classifier classify-samples-ncv \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column Preservation \\\n",
    "--p-estimator RandomForestClassifier \\\n",
    "--p-n-estimators 1000 \\\n",
    "--p-random-state 123 \\\n",
    "--o-predictions dada2/core-metrics-results-rar11987/Preservation-predictions-RFncv.qza \\\n",
    "--o-probabilities dada2/core-metrics-results-rar11987/Preservation-probabilities-RFncv.qza \\\n",
    "--o-feature-importance dada2/core-metrics-results-rar11987/Preservation-feature-importance-RFncv.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/Preservation-confusion-matrix-RFncv.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/Preservation-feature-importance-RFncv.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#examine differences in distinguishing classes\n",
    "!qiime sample-classifier confusion-matrix \\\n",
    "--i-predictions dada2/core-metrics-results-rar11987/Preservation-predictions-RFncv.qza \\\n",
    "--i-probabilities dada2/core-metrics-results-rar11987/Preservation-probabilities-RFncv.qza \\\n",
    "--m-truth-file metadata-extra202412.txt \\\n",
    "--m-truth-column Preservation \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/Preservation-confusion-matrix-RFncv.qzv\n",
    "\n",
    "#examine feature weights\n",
    "!qiime metadata tabulate \\\n",
    "--m-input-file dada2/core-metrics-results-rar11987/Preservation-feature-importance-RFncv.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/Preservation-feature-importance-RFncv.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Predict samples from different subjects based on their Subject ID's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cluster/project/bokulich/alavrinienko/_conda_envs/qiime2-amplicon-2024.10/lib/python3.10/site-packages/q2_sample_classifier/utilities.py:715: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  imp = imp.groupby(imp.columns, axis=1).mean()\n",
      "Estimator Accuracy: 0.9882352941176471 ± 0.015686274509803914\n",
      "\u001b[32mSaved SampleData[ClassifierPredictions] to: dada2/core-metrics-results-rar11987/SubjectID-predictions-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Importance] to: dada2/core-metrics-results-rar11987/SubjectID-feature-importance-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[Probabilities] to: dada2/core-metrics-results-rar11987/SubjectID-probabilities-RFncv.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#use filtered rarefied data for sample classifier\n",
    "!qiime sample-classifier classify-samples-ncv \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column SubjectID \\\n",
    "--p-estimator RandomForestClassifier \\\n",
    "--p-n-estimators 1000 \\\n",
    "--p-random-state 123 \\\n",
    "--o-predictions dada2/core-metrics-results-rar11987/SubjectID-predictions-RFncv.qza \\\n",
    "--o-probabilities dada2/core-metrics-results-rar11987/SubjectID-probabilities-RFncv.qza \\\n",
    "--o-feature-importance dada2/core-metrics-results-rar11987/SubjectID-feature-importance-RFncv.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/SubjectID-confusion-matrix-RFncv.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/SubjectID-feature-importance-RFncv.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#examine differences in distinguishing classes\n",
    "!qiime sample-classifier confusion-matrix \\\n",
    "--i-predictions dada2/core-metrics-results-rar11987/SubjectID-predictions-RFncv.qza \\\n",
    "--i-probabilities dada2/core-metrics-results-rar11987/SubjectID-probabilities-RFncv.qza \\\n",
    "--m-truth-file metadata-extra202412.txt \\\n",
    "--m-truth-column SubjectID \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/SubjectID-confusion-matrix-RFncv.qzv\n",
    "\n",
    "#examine feature weights\n",
    "!qiime metadata tabulate \\\n",
    "--m-input-file dada2/core-metrics-results-rar11987/SubjectID-feature-importance-RFncv.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/SubjectID-feature-importance-RFncv.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine if the prediction accuracy for the SubjectID is influenced by the presence of 'directly-plated' fresh samples (i.e., check if the SubjectID can still be accurately predicted only by using the microbial composition data in the samples recovered after the long-term storage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#filter samples that were directly plated - retain only samples recovered after long-term storage\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-where \"[Group]='Direct-Plating'\" \\\n",
    "--p-exclude-ids True \\\n",
    "--o-filtered-table dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qza\n",
    "\n",
    "#summarize and validate filtered rarefied data\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Re-run the sample classifier usign these filtered data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cluster/project/bokulich/alavrinienko/_conda_envs/qiime2-amplicon-2024.10/lib/python3.10/site-packages/q2_sample_classifier/utilities.py:715: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  imp = imp.groupby(imp.columns, axis=1).mean()\n",
      "Estimator Accuracy: 1.0 ± 0.0\n",
      "\u001b[32mSaved SampleData[ClassifierPredictions] to: dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-predictions-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Importance] to: dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-feature-importance-RFncv.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[Probabilities] to: dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-probabilities-RFncv.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#use filtered rarefied data for sample classifier\n",
    "!qiime sample-classifier classify-samples-ncv \\\n",
    "--i-table dada2/core-metrics-results-rar11987/rarefied_nofresh_nodirectplating_table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--m-metadata-column SubjectID \\\n",
    "--p-estimator RandomForestClassifier \\\n",
    "--p-n-estimators 1000 \\\n",
    "--p-random-state 123 \\\n",
    "--o-predictions dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-predictions-RFncv.qza \\\n",
    "--o-probabilities dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-probabilities-RFncv.qza \\\n",
    "--o-feature-importance dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-feature-importance-RFncv.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-confusion-matrix-RFncv.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-feature-importance-RFncv.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#examine differences in distinguishing classes\n",
    "!qiime sample-classifier confusion-matrix \\\n",
    "--i-predictions dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-predictions-RFncv.qza \\\n",
    "--i-probabilities dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-probabilities-RFncv.qza \\\n",
    "--m-truth-file metadata-extra202412.txt \\\n",
    "--m-truth-column SubjectID \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-confusion-matrix-RFncv.qzv\n",
    "\n",
    "#examine feature weights\n",
    "!qiime metadata tabulate \\\n",
    "--m-input-file dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-feature-importance-RFncv.qza \\\n",
    "--o-visualization dada2/core-metrics-results-rar11987/SubjectID-nodirectplating-feature-importance-RFncv.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential abundance testing\n",
    "\n",
    "Use ANCOM-BC implemented in the q2-composition plugin for testing differentially abundant features across study groups. First, the feature data have to be prepared for the differential abundance (DA) testing by filtering to subset relevant samples (note that the unrarefied data are used for the DA testing)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run DA tests using samples from adults"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir dada2/ancombc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/adults-plated-feature-table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/adults-plated-feature-table.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#keep only samples from ADULTS + remove samples that were sequenced fresh (only plated samples are considered here)\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-where \"[HostAgeGroup]='Adult' AND [State]='Pooled'\" \\\n",
    "--o-filtered-table dada2/ancombc/adults-plated-feature-table.qza\n",
    "\n",
    "#check the filtering outcome - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/adults-plated-feature-table.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/adults-plated-feature-table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/adults-plated-feature-table-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/adults-plated-feature-table-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#collapse the filtered feature data at the genus level\n",
    "!qiime taxa collapse \\\n",
    "--i-table dada2/ancombc/adults-plated-feature-table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--p-level 6 \\\n",
    "--o-collapsed-table dada2/ancombc/adults-plated-feature-table-taxL6.qza\n",
    "\n",
    "#check the count of genera retained in this subset - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/adults-plated-feature-table-taxL6.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/adults-plated-feature-table-taxL6.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[DifferentialAbundance] to: dada2/ancombc/adults-plated-ancombc-diff-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/adults-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/da-barplot-adults-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition ancombc \\\n",
    "--i-table dada2/ancombc/adults-plated-feature-table-taxL6.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-formula 'Preservation' \\\n",
    "--p-reference-levels \"Preservation::Direct-Plating\" \\\n",
    "--o-differentials dada2/ancombc/adults-plated-ancombc-diff-taxL6.qza\n",
    "\n",
    "!qiime composition tabulate \\\n",
    "--i-data dada2/ancombc/adults-plated-ancombc-diff-taxL6.qza \\\n",
    "--o-visualization dada2/ancombc/adults-plated-ancombc-diff-taxL6.qzv\n",
    "\n",
    "!qiime composition da-barplot \\\n",
    "--i-data dada2/ancombc/adults-plated-ancombc-diff-taxL6.qza \\\n",
    "--p-significance-threshold 0.001 \\\n",
    "--p-level-delimiter ';' \\\n",
    "--o-visualization dada2/ancombc/da-barplot-adults-plated-ancombc-diff-taxL6.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run DA tests using samples from children"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/children-plated-feature-table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/children-plated-feature-table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/children-plated-feature-table-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/children-plated-feature-table-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#keep only samples from CHILDREN + remove samples that were sequenced fresh\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-where \"[HostAgeGroup]='Child' AND [State]='Pooled'\" \\\n",
    "--o-filtered-table dada2/ancombc/children-plated-feature-table.qza\n",
    "\n",
    "#check the filtering outcome - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/children-plated-feature-table.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/children-plated-feature-table.qzv\n",
    "\n",
    "#collapse the filtered feature data at the genus level\n",
    "!qiime taxa collapse \\\n",
    "--i-table dada2/ancombc/children-plated-feature-table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--p-level 6 \\\n",
    "--o-collapsed-table dada2/ancombc/children-plated-feature-table-taxL6.qza\n",
    "\n",
    "#check the count of genera retained in this subset - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/children-plated-feature-table-taxL6.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/children-plated-feature-table-taxL6.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[DifferentialAbundance] to: dada2/ancombc/children-plated-ancombc-diff-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/children-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/da-barplot-children-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition ancombc \\\n",
    "--i-table dada2/ancombc/children-plated-feature-table-taxL6.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-formula 'Preservation' \\\n",
    "--p-reference-levels \"Preservation::Direct-Plating\" \\\n",
    "--o-differentials dada2/ancombc/children-plated-ancombc-diff-taxL6.qza\n",
    "\n",
    "!qiime composition tabulate \\\n",
    "--i-data dada2/ancombc/children-plated-ancombc-diff-taxL6.qza \\\n",
    "--o-visualization dada2/ancombc/children-plated-ancombc-diff-taxL6.qzv\n",
    "\n",
    "!qiime composition da-barplot \\\n",
    "--i-data dada2/ancombc/children-plated-ancombc-diff-taxL6.qza \\\n",
    "--p-significance-threshold 0.001 \\\n",
    "--p-level-delimiter ';' \\\n",
    "--o-visualization dada2/ancombc/da-barplot-children-plated-ancombc-diff-taxL6.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run DA tests using samples from infants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/infants-plated-feature-table.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/infants-plated-feature-table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: dada2/ancombc/infants-plated-feature-table-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/infants-plated-feature-table-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#keep only samples from INFANTS + remove samples that were sequenced fresh\n",
    "!qiime feature-table filter-samples \\\n",
    "--i-table dada2/merged-clean-taxfilt-feature-table.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-where \"[HostAgeGroup]='Infant' AND [State]='Pooled'\" \\\n",
    "--o-filtered-table dada2/ancombc/infants-plated-feature-table.qza\n",
    "\n",
    "#check the filtering outcome - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/infants-plated-feature-table.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/infants-plated-feature-table.qzv\n",
    "\n",
    "#collapse the filtered feature data at the genus level\n",
    "!qiime taxa collapse \\\n",
    "--i-table dada2/ancombc/infants-plated-feature-table.qza \\\n",
    "--i-taxonomy dada2/val-exp-merged-clean-silva-taxonomy.qza \\\n",
    "--p-level 6 \\\n",
    "--o-collapsed-table dada2/ancombc/infants-plated-feature-table-taxL6.qza\n",
    "\n",
    "#check the count of genera retained in this subset - OK\n",
    "!qiime feature-table summarize \\\n",
    "--i-table dada2/ancombc/infants-plated-feature-table-taxL6.qza \\\n",
    "--m-sample-metadata-file metadata-extra202412.txt \\\n",
    "--o-visualization dada2/ancombc/infants-plated-feature-table-taxL6.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[DifferentialAbundance] to: dada2/ancombc/infants-plated-ancombc-diff-taxL6.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/infants-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: dada2/ancombc/da-barplot-infants-plated-ancombc-diff-taxL6.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition ancombc \\\n",
    "--i-table dada2/ancombc/infants-plated-feature-table-taxL6.qza \\\n",
    "--m-metadata-file metadata-extra202412.txt \\\n",
    "--p-formula 'Preservation' \\\n",
    "--p-reference-levels \"Preservation::Direct-Plating\" \\\n",
    "--o-differentials dada2/ancombc/infants-plated-ancombc-diff-taxL6.qza\n",
    "\n",
    "!qiime composition tabulate \\\n",
    "--i-data dada2/ancombc/infants-plated-ancombc-diff-taxL6.qza \\\n",
    "--o-visualization dada2/ancombc/infants-plated-ancombc-diff-taxL6.qzv\n",
    "\n",
    "!qiime composition da-barplot \\\n",
    "--i-data dada2/ancombc/infants-plated-ancombc-diff-taxL6.qza \\\n",
    "--p-significance-threshold 0.001 \\\n",
    "--p-level-delimiter ';' \\\n",
    "--o-visualization dada2/ancombc/da-barplot-infants-plated-ancombc-diff-taxL6.qzv"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "15b467d9886749254aaced2c24db3236060557b4935f56902b40660c3177ee38"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
